### PR TITLE
Move parse regex logic to template renderer

### PR DESCRIPTION
**Pull request by vicancy**

### CODE REVIEW

**Comment by ansyral** [comment]
>         private static readonly Regex IsRegexPatternRegex = new Regex(@"^\s*/(.*)/\s*$", RegexOptions.Compiled);

remove?

---
Refers to: src/Microsoft.DocAsCode.Build.Engine/Template.cs:19 in 5b68d62. [](commit_id = 5b68d62e3d43cdee2a537823e0a93caecc75ab49, deletion_comment = False)

**Comment by ansyral** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/TemplateRenderers/LiquidTemplateRenderer.cs [hunk 1]
```diff
@@ -7,32 +7,49 @@ namespace Microsoft.DocAsCode.Build.Engine
...
-        private static object _locker = new object();
-        private readonly DotLiquid.Template _template = null;
-        public static LiquidTemplateRenderer Create(ResourceCollection resourceProvider, string template)
+        private static readonly object _locker = new object();
+
+        private readonly DotLiquid.Template _template;
+
+        public static LiquidTemplateRenderer Create(ResourceCollection resourceProvider, string template, string templateName)
         {
             if (template == null) throw new ArgumentNullException(nameof(template));
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateRenderers/LiquidTemplateRenderer.cs]: >template [](start = 16, length = 8)

minor: check templateName

File: src/Microsoft.DocAsCode.Build.Engine/TemplateRenderers/MustacheTemplateRenderer.cs [hunk 2]
```diff
@@ -6,21 +6,32 @@ namespace Microsoft.DocAsCode.Build.Engine
     using System;
     using System.Collections.Generic;
     using System.IO;
+    using System.Linq;
     using System.Text.RegularExpressions;
 
     internal class MustacheTemplateRenderer : ITemplateRenderer
     {
         private static readonly Regex IncludeRegex = new Regex(@"{{\s*!\s*include\s*\(:?(:?['""]?)\s*(?<file>(.+?))\1\s*\)\s*}}", RegexOptions.Compiled | RegexOptions.IgnoreCase);
+        private static readonly Regex IsRegexPatternRegex = new Regex(@"^\s*/(.*)/\s*$", RegexOptions.Compiled);
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateRenderers/MustacheTemplateRenderer.cs]: >IsRegexPatternRegex [](start = 38, length = 19)

no use?

**Comment by superyyrrzz** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/TemplateResourceInfo.cs [hunk 3]
```diff
@@ -6,14 +6,35 @@ namespace Microsoft.DocAsCode.Build.Engine
...
+            if (ReferenceEquals(this, obj))
+            {
+                return true;
+            }
+
+            var target = obj as TemplateResourceInfo;
+            if (target == null)
+            {
+                return false;
+            }
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateResourceInfo.cs]: could put this logic first and remove line 15?

File: src/Microsoft.DocAsCode.Build.Engine/Template.cs [hunk 4]
```diff
@@ -235,15 +217,15 @@ private static ITemplatePreprocessor CreatePreprocessor(ResourceCollection resou
             return new TemplateJintPreprocessor(resourceCollection, scriptResource, context);
         }
 
-        private static ITemplateRenderer CreateRenderer(ResourceCollection resourceCollection, TemplateRendererResource templateResource)
+        private static ITemplateRenderer CreateRenderer(ResourceCollection resourceCollection, TemplateRendererResource templateResource, string templateName)
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/Template.cs]: >templateName [](start = 145, length = 12)

could the templateName fetch from templateResource.ResouceName?

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/Template.cs]: if can, how about passing TemplateRendererResource as param name?

---
In reply to: [109278891](https://github.com/dotnet/docfx/pull/1513#discussion_r109278891) [](ancestors = 109278891)

File: src/Microsoft.DocAsCode.Build.Engine/TemplateRenderers/LiquidTemplateRenderer.cs [hunk 5]
```diff
@@ -7,32 +7,57 @@ namespace Microsoft.DocAsCode.Build.Engine
...
+                throw new ArgumentNullException(nameof(templateName));
+            }
+
+            // Guarantee that each time returns a new renderer
+            // As Dependency is a globally shared object, allow one entry at a time
             lock (_locker)
             {
                 DotLiquid.Template.RegisterTag<Dependency>("ref");
                 Dependency.PopDependencies();
                 var liquidTemplate = DotLiquid.Template.Parse(template);
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateRenderers/LiquidTemplateRenderer.cs]: >Parse [](start = 56, length = 5)

when parse and pop dependencies, the lock of dependency will lock again?

Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateRenderers/LiquidTemplateRenderer.cs]: recursive lock, see [https://en.wikipedia.org/wiki/Reentrant_mutex](https://en.wikipedia.org/wiki/Reentrant_mutex)

---
In reply to: [109279261](https://github.com/dotnet/docfx/pull/1513#discussion_r109279261) [](ancestors = 109279261)

File: src/Microsoft.DocAsCode.Build.Engine/TemplateRenderers/LiquidTemplateRenderer.cs [hunk 6]
```diff
@@ -7,32 +7,57 @@ namespace Microsoft.DocAsCode.Build.Engine
...
         {
             _template = liquidTemplate;
             Raw = template;
-            Dependencies = dependencies;
+            Dependencies = ParseDependencies(templateName, resource, dependencies).ToList();
+        }
+
+        private IEnumerable<string> ParseDependencies(string templateName, ResourceCollection resource, IEnumerable<string> raw)
+        {
+            foreach (var item in raw)
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateRenderers/LiquidTemplateRenderer.cs]: from ... from ... select ...?

**Comment by vwxyzh** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/TemplateRenderers/LiquidTemplateRenderer.cs (hunk 5 above)
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateRenderers/LiquidTemplateRenderer.cs]: _locker is a static lock, why there is another lock?

---
In reply to: [109279261](https://github.com/dotnet/docfx/pull/1513#discussion_r109279261) [](ancestors = 109279261)

File: src/Microsoft.DocAsCode.Build.Engine/Template.cs (hunk 4 above)
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/Template.cs]: sounds cleaner

---
In reply to: [109278925](https://github.com/dotnet/docfx/pull/1513#discussion_r109278925) [](ancestors = 109278925,109278891)

**Comment by hellosnow** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/Template.cs (hunk 4 above)
Comment (928PJY) [inline src/Microsoft.DocAsCode.Build.Engine/Template.cs]: >, string templateName [](start = 136, length = 21)

remove?

**Pull request merged**